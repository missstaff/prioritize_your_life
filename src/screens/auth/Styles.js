import { ScaledSheet, moderateScale, scale, verticalScale } from "react-native-size-matters";
import { APP_COLORS } from "../../utility/constants";

export const styles = ScaledSheet.create({
    container: {
        alignItems: "center",
        backgroundColor: APP_COLORS.primary,
        flex: 1,
        justifyContent: "center",
    },
    title: {
        color: APP_COLORS.secondary,
        fontSize: scale(50),
        fontWeight: "bold",
        marginBottom: verticalScale(40),
    },
    inputView: {
        backgroundColor: APP_COLORS.quaternary,
        borderRadius: moderateScale(25),
        height: scale(50),
        justifyContent: "center",
        marginBottom: verticalScale(20),
        padding: scale(20),
        width: "80%",
    },
    inputText: {
        color: APP_COLORS.white,
        height: scale(50),
    },
    forgotAndSignUpText: {
        color: APP_COLORS.white,
        fontSize: scale(11),
    },
    loginBtn: {
        alignItems: "center",
        backgroundColor: APP_COLORS.secondary,
        borderRadius: moderateScale(25),
        height: scale(50),
        justifyContent: "center",
        marginBottom: verticalScale(10),
        marginTop: verticalScale(40),
        width: "80%",
    },
});