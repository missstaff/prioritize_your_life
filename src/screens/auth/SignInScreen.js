import React, { useState } from "react";
import { Text, TouchableOpacity, View } from "react-native";
import { useNavigation } from "@react-navigation/native";
import AppTextInput from "../../components/ui/AppTextInput";
import { validateTextInput } from "../../utility/auth-utilities";
import { displayToast } from "../../utility/utilities";
import { APP_COLORS } from "../../utility/constants";
import { styles } from "./Styles";
import { getFireApp } from "../../../getFireApp";
const firebase = getFireApp();


const SignInScreen = () => {

    const navigation = useNavigation();

    const [state, setState] = useState({
        email: "",
        password: "",
    });

    const onPressForgotPassword = () => {
        navigation.navigate("PasswordResetScreen");
    };

    const onPressSignUp = () => {
        navigation.navigate("SignUp");
    };

    const onPressLogin = async () => {

        let user;
        let isValidEmail = false;
        let isValidPassword = false;


        isValidEmail = validateTextInput({
            condition: state.email,
            id: "invalid_email",
            type: "error",
        });

        isValidPassword = validateTextInput({
            condition: state.password,
            id: "password_required",
            type: "error",
        });


        if (!isValidEmail || !isValidPassword) {
            return;
        }


        try {
            user = await firebase.auth().signInWithEmailAndPassword(state.email, state.password);
        } catch (error) {
            console.log(`Error: ${error.message}\n${error.stack}`);
            displayToast("login_failure", "error");
            return;
        }

        try {
            firebase.firestore().collection("users").doc(user.user.uid).update({
                last_login: new Date(),
            });
        } catch (error) {
            console.log(`Error: ${error.message}\n${error.stack}`);
        }

    };


    return (
        <View style={styles.container}>
            <Text style={styles.title}>
                Login
            </Text>
            <View style={styles.inputView}>
                <AppTextInput
                    style={styles.inputText}
                    placeholder="Email"
                    placeholderTextColor={APP_COLORS.triodenary}
                    onChangeText={(text) => setState({ ...state, email: text })} />
            </View>
            <View style={styles.inputView}>
                <AppTextInput
                    style={styles.inputText}
                    secureTextEntry
                    placeholder="Password"
                    placeholderTextColor={APP_COLORS.triodenary}
                    onChangeText={(text) => setState({ ...state, password: text })} />
            </View>

            <TouchableOpacity
                accessibilityRole="link"
                accessibilityLabel="Forgot password link"
                onPress={onPressForgotPassword}>
                <Text style={styles.forgotAndSignUpText}>
                    Forgot Password?
                </Text>
            </TouchableOpacity>
            <TouchableOpacity
                accessibilityRole="button"
                accessibilityLabel="Login button"
                onPress={onPressLogin}
                style={styles.loginBtn}>
                <Text style={styles.loginText}>
                    LOGIN
                </Text>
            </TouchableOpacity>
            <TouchableOpacity
                accessibilityRole="link"
                accessibilityLabel="Signup link"
                onPress={onPressSignUp}>
                <Text style={styles.forgotAndSignUpText}>
                    Signup
                </Text>
            </TouchableOpacity>
        </View>
    );
};

export default SignInScreen;