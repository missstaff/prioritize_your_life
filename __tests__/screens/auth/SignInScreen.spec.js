import renderer from "react-test-renderer";
import { fireEvent, render, waitFor } from '@testing-library/react-native';
import * as authModule from '@react-native-firebase/auth';
import { useNavigation } from "@react-navigation/native";
import SignInScreen from "../../../src/screens/auth/SignInScreen";


describe("SignInScreen", () => {

    afterAll(() => {
        jest.clearAllMocks();
    });

    it('renders correctly', () => {
        const { getByText, getByPlaceholderText } = render(<SignInScreen />);
        
        // Test rendering of components
        expect(getByText('Login')).toBeTruthy();
        expect(getByPlaceholderText('Email')).toBeTruthy();
        expect(getByPlaceholderText('Password')).toBeTruthy();
      });
    
      it('navigates to password reset screen', () => {
        const { getByText } = render(<SignInScreen />);
        const forgotPasswordLink = getByText('Forgot Password?');
        
        fireEvent.press(forgotPasswordLink);
        
        expect(useNavigation().navigate).toHaveBeenCalledWith('PasswordResetScreen');
      });

      it('navigates to signup screen', () => {
        const { getByText } = render(<SignInScreen />);
        const signupLink = getByText('Signup');
  
        fireEvent.press(signupLink);
        
        expect(useNavigation().navigate).toHaveBeenCalledWith('SignUp');
      });
    
      it('logs in with valid credentials', async() => {
        const { getByText, getByPlaceholderText } = render(<SignInScreen />);
        const loginButton = getByText('LOGIN');
    
        fireEvent.changeText(getByPlaceholderText('Email'), 'shawnastaff@gmail.com');
        fireEvent.changeText(getByPlaceholderText('Password'), 'Abc123!');
    
        fireEvent.press(loginButton);
        
        waitFor(() => {
            expect(authModule.signInWithEmailAndPassword)
              .toHaveBeenCalledWith('shawnastaff@gmail.com', 'Abc123!');
          });
      });
    
    
      it('matches snapshot', () => {
        const tree = renderer.create(<SignInScreen />).toJSON();
        expect(tree).toMatchSnapshot();
      });
});