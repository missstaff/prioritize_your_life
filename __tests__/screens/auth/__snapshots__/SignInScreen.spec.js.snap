// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInScreen matches snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#4FD3DA",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      {
        "color": "#FFC857",
        "fontSize": 107.14285714285714,
        "fontWeight": "bold",
        "marginBottom": 78.47058823529412,
      }
    }
  >
    Login
  </Text>
  <View
    style={
      {
        "backgroundColor": "#3AB4BA",
        "borderRadius": 39.285714285714285,
        "height": 107.14285714285714,
        "justifyContent": "center",
        "marginBottom": 39.23529411764706,
        "padding": 42.857142857142854,
        "width": "80%",
      }
    }
  >
    <View>
      <TextInput
        onChangeText={[Function]}
        placeholder="Email"
        placeholderTextColor="#003f5c"
        style={
          {
            "color": "#FFFFFF",
            "height": 107.14285714285714,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      {
        "backgroundColor": "#3AB4BA",
        "borderRadius": 39.285714285714285,
        "height": 107.14285714285714,
        "justifyContent": "center",
        "marginBottom": 39.23529411764706,
        "padding": 42.857142857142854,
        "width": "80%",
      }
    }
  >
    <View>
      <TextInput
        onChangeText={[Function]}
        placeholder="Password"
        placeholderTextColor="#003f5c"
        style={
          {
            "color": "#FFFFFF",
            "height": 107.14285714285714,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityLabel="Forgot password link"
    accessibilityRole="link"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontSize": 23.57142857142857,
        }
      }
    >
      Forgot Password?
    </Text>
  </View>
  <View
    accessibilityLabel="Login button"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFC857",
        "borderRadius": 39.285714285714285,
        "height": 107.14285714285714,
        "justifyContent": "center",
        "marginBottom": 19.61764705882353,
        "marginTop": 78.47058823529412,
        "opacity": 1,
        "width": "80%",
      }
    }
  >
    <Text>
      LOGIN
    </Text>
  </View>
  <View
    accessibilityLabel="Signup link"
    accessibilityRole="link"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontSize": 23.57142857142857,
        }
      }
    >
      Signup
    </Text>
  </View>
</View>
`;
